File: frontend/src/features/attendance/api/attendance.ts
import { apiClient } from '@/api/client';

export interface AttendanceStats {
    ok: number;
    ng: number;
    pending: number;
    total: number;
}

export interface AttendanceEventResponse {
    id: string;
    project_id: string;
    title: string;
    schedule_date: string | null;
    deadline: string | null;
    completed: boolean;
    created_at: string;
    stats: AttendanceStats;
}

export interface AttendanceTargetResponse {
    user_id: string;
    display_name: string | null;
    discord_username: string;
    status: 'pending' | 'ok' | 'ng';
}

export interface AttendanceEventDetailResponse extends AttendanceEventResponse {
    targets: AttendanceTargetResponse[];
}

export const attendanceApi = {
    // 出欠確認一覧を取得
    getAttendanceEvents: async (projectId: string): Promise<AttendanceEventResponse[]> => {
        const response = await apiClient.get<AttendanceEventResponse[]>(`/projects/${projectId}/attendance`);
        return response.data;
    },

    // 出欠確認詳細を取得
    getAttendanceEvent: async (projectId: string, eventId: string): Promise<AttendanceEventDetailResponse> => {
        const response = await apiClient.get<AttendanceEventDetailResponse>(`/projects/${projectId}/attendance/${eventId}`);
        return response.data;
    },

    // 出欠確認を外部連携用JSONとして出力
    exportAttendanceEvent: async (projectId: string, eventId: string): Promise<{ blob: Blob; filename: string | null }> => {
        const response = await apiClient.get<Blob>(`/projects/${projectId}/attendance/${eventId}/export`, {
            responseType: 'blob',
        });
        const disposition = response.headers['content-disposition'];
        const filenameMatch = typeof disposition === 'string'
            ? disposition.match(/filename="([^"]+)"/)
            : null;

        return {
            blob: response.data,
            filename: filenameMatch?.[1] ?? null,
        };
    },

    // Pendingユーザーにリマインダー送信
    remindPendingUsers: async (projectId: string, eventId: string): Promise<{ message: string }> => {
        const response = await apiClient.post<{ message: string }>(`/projects/${projectId}/attendance/${eventId}/remind-pending`);
        return response.data;
    },

    // 自分の出欠確認ステータスを更新
    updateMyAttendanceStatus: async (
        projectId: string,
        eventId: string,
        status: 'ok' | 'ng' | 'pending'
    ): Promise<AttendanceEventDetailResponse> => {
        const response = await apiClient.patch<AttendanceEventDetailResponse>(
            `/projects/${projectId}/attendance/${eventId}/my-status`,
            { status }
        );
        return response.data;
    },
};

File: backend/src/schemas/attendance.py
"""出席確認スキーマ."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class AttendanceStats(BaseModel):
    """出席確認統計."""

    ok: int
    ng: int
    pending: int
    total: int


class AttendanceEventResponse(BaseModel):
    """出席確認イベントレスポンス."""

    id: UUID
    project_id: UUID
    rehearsal_id: UUID | None
    title: str
    schedule_date: datetime | None
    deadline: datetime | None
    completed: bool
    created_at: datetime
    stats: AttendanceStats


class AttendanceTargetResponse(BaseModel):
    """出席確認ターゲットレスポンス."""

    user_id: UUID
    display_name: str | None
    discord_username: str
    status: str  # "ok", "ng", "pending"


class AttendanceEventDetailResponse(BaseModel):
    """出席確認イベント詳細レスポンス."""

    id: UUID
    project_id: UUID
    rehearsal_id: UUID | None
    title: str
    schedule_date: datetime | None
    deadline: datetime | None
    completed: bool
    created_at: datetime
    stats: AttendanceStats
    targets: list[AttendanceTargetResponse]


class AttendanceExportTarget(BaseModel):
    """外部連携用出欠対象者."""

    name: str
    status: str


class AttendanceExportResponse(BaseModel):
    """外部連携用出欠JSONレスポンス."""

    model_config = ConfigDict(populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion")
    source: str
    event_name: str = Field(alias="eventName")
    generated_at: str = Field(alias="generatedAt")
    attendances: list[AttendanceExportTarget]


class AttendanceTargetUpdate(BaseModel):
    """出席確認ターゲット更新."""

    remove_user_ids: list[str] = []
    add_user_ids: list[str] = []


class AttendanceStatusUpdate(BaseModel):
    """自分の出席確認ステータス更新."""

    status: str  # "ok", "ng", "pending"
